Feature/312 json validation

zehnm opened the pull request:
JSON validation with [valijson](https://github.com/tristanpenman/valijson) for config.json, hardware.json and the websocket API for incoming configuration.

Details:
- Currently any validation errors are logged and the app continues to start up.
  - main.cpp is prepared to be enhanced with an error screen
- Using valijson's Qt adapter, so the document doesn't need to be loaded and parsed twice.
- No further dependencies on other libraries.
- Easy integration through header files only.
- Initial support for JSON schema v6 and v7 by using dev branch of valijson.

This closes #312

zehnm committed:
#312 JSON schema validation for config.json

Initial JSON schema validation for the main configuration with valijson.
Not yet implemented are:
- hardware.json validation
- configuration received over websocket API

zehnm committed:
#312 JSON schema validation for hardware.json

Initial JSON schema validation for the hardware configuration file.

zehnm committed:
#312 JSON schema validation for websocket API

- Initial JSON schema validation for incoming configuration through the
  websocket API.
- Validate configuration before writing to file.

zehnm committed:
Merge branch 'dev' into feature/312_json_validation

zehnm committed:
#312 Quick fix for invalid configuration

Use best effort for now and simply log schema validation errors.
An empty configuration will fail the app to load.

zehnm committed:
#312 config JSON schema improvements

- Valid default config.json
- Initial cleanup of config-schema.json

martonborzak commented on sources/config.h:
```diff
@@ -38,6 +38,8 @@ class Config : public QObject, ConfigInterface
     Q_INTERFACES(ConfigInterface)
 
 public:
+    Q_PROPERTY (bool             valid              READ isValid                                    CONSTANT)
```
Missing NOTIFY. That is needed so we can connect to a signal from QML.

zehnm replied:
Do you really need to expose all properties to QML?
I would rather remove as many properties as possible if they are not used in the UI at the moment. Configuration setting & validation happens only at app startup and in the YIO API (which is a cpp class). Or did I miss something?

martonborzak replied:
You might want to display an error notification on the remote. That's the reason I think we should have at least one signal.

zehnm replied:
At the time the configuration is loaded there's no UI. For that we need something new and the error is already propagated all the way to the application loading in main.cpp.
For the websocket API there's a TODO for an error response. In my opinion it doesn't make sense to display an error on the remote when using the web-configurator:  
In case the web-configurator sends invalid config data, it is not stored in the config object. I.e. the old configuration is still active on the remote.
We should keep the config & jsonfile classes as lean as possible and only expose QML properties and invokable methods which are really required at the moment. This makes refactoring much easier.

martonborzak replied:
Right. Makes sense. Agree about the web configurator. The error should be on the web interface.

zehnm replied:
Can I remove all unused properties or would you like to keep them?

martonborzak commented on sources/config.h:
```diff
@@ -38,6 +38,8 @@ class Config : public QObject, ConfigInterface
 
 public:
+    Q_PROPERTY (bool             valid              READ isValid                                    CONSTANT)
+    Q_PROPERTY (QString          error              READ getError                                   CONSTANT)
```
Missing NOTIFY. That is needed so we can connect to a signal from QML.

martonborzak commented on sources/jsonfile.cpp:
```diff
@@ -87,23 +119,101 @@ bool JsonFile::write(const QVariantMap &data)
+        return true;
+    }
+
+    // TODO in case there are too many read or write calls we should probably cache the schema document
```
There are a lot of calls at the moment. Caching is a good idea :)

martonborzak commented on sources/jsonfile.h:
```diff
@@ -30,27 +31,41 @@
+    Q_PROPERTY(bool        writeable      READ writeable)
+    Q_PROPERTY(bool        readable       READ readable)
+    Q_PROPERTY(qint64      size           READ size)
+    Q_PROPERTY(bool        valid          READ isValid                                CONSTANT)
```
Missing NOTIFY. That is needed so we can connect to a signal from QML.

martonborzak commented on sources/jsonfile.h:
```diff
@@ -30,27 +31,41 @@
+    Q_PROPERTY(bool        readable       READ readable)
+    Q_PROPERTY(qint64      size           READ size)
+    Q_PROPERTY(bool        valid          READ isValid                                CONSTANT)
+    Q_PROPERTY(QString     error          READ error                                  CONSTANT)
```
Missing NOTIFY. That is needed so we can connect to a signal from QML.

martonborzak reviewed (changes requested):
I think it's good, only the missing NOTIFY signals can be an issue if we want to communicate with QML.

martonborzak commented:
Feel free to remove them! Thanks

> On 27 Dec 2019, at 12.15, Markus Zehnder <notifications@github.com> wrote:
> 
> @zehnm commented on this pull request.
> 
> In sources/config.h <https://github.com/YIO-Remote/remote-software/pull/374#discussion_r361642026>:
> 
> > @@ -38,6 +38,8 @@ class Config : public QObject, ConfigInterface
>      Q_INTERFACES(ConfigInterface)
>  
>  public:
> +    Q_PROPERTY (bool             valid              READ isValid                                    CONSTANT)
> Can I remove all unused properties or would you like to keep them?
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/YIO-Remote/remote-software/pull/374?email_source=notifications&email_token=AFQRIUB42WRRITNWMA377GDQ2XPTXA5CNFSM4J7Y4GSKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCQJEWIQ#discussion_r361642026>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AFQRIUCBOHKCHPZMS2AAWV3Q2XPTXANCNFSM4J7Y4GSA>.
>

zehnm committed:
Merge branch 'dev' into feature/312_json_validation

zehnm committed:
Fixed cpplint issues

zehnm committed:
Further cpplint fixes

zehnm committed:
Use cpplint root for proper header guards

 Removed paths: action was triggered twice together with branches

zehnm committed:
Using NOLINT markers for certain issues

zehnm committed:
Stupid UTF-8 BOM made cpplint go crazy

zehnm committed:
Restored push paths for cpp files

Since this branch is already in a pull request it's no wonder two checks are
executed. Duh!

zehnm committed:
GitHub Action for config schema validation

zehnm committed:
Formatted with clang Google style
